modules/library_borrow/library.py

```diff
@@ -145,38 +145,36 @@ class Book(metaclass=PoolMeta):
 
     @classmethod
     def getter_is_available(cls, books, name):
-        pool = Pool()
-        checkout = pool.get('library.user.checkout').__table__()
-        exemplary = pool.get('library.book.exemplary').__table__()
-        book = cls.__table__()
-        result = {x.id: False for x in books}
-        cursor = Transaction().connection.cursor()
-        cursor.execute(*book.join(exemplary,
-                condition=(exemplary.book == book.id)
-                ).join(checkout, 'LEFT OUTER',
-                condition=(exemplary.id == checkout.exemplary)
-                ).select(book.id,
-                where=(checkout.return_date != Null) | (checkout.id == Null)))
-        for book_id, in cursor.fetchall():
-            result[book_id] = True
+        Book = Pool().get('library.book')
+        books_ids = [x.id for x in books]
+        books = Book.search([('id', 'in', books_ids)])
+        result = {}
+        for book in books:
+            result[book.id] = False
+            for exemplary in book.exemplaries:
+                if exemplary.is_available:
+                    result[book.id] = True
+                    break
+                else:
+                    continue
         return result
 
     @classmethod
     def search_is_available(cls, name, clause):
         _, operator, value = clause
         if operator == '!=':
             value = not value
-        pool = Pool()
-        checkout = pool.get('library.user.checkout').__table__()
-        exemplary = pool.get('library.book.exemplary').__table__()
-        book = cls.__table__()
-        query = book.join(exemplary,
-            condition=(exemplary.book == book.id)
-            ).join(checkout, 'LEFT OUTER',
-            condition=(exemplary.id == checkout.exemplary)
-            ).select(book.id,
-            where=(checkout.return_date != Null) | (checkout.id == Null))
-        return [('id', 'in' if value else 'not in', query)]
+        Book = Pool().get('library.book')
+        books = Book.search([])
```

Comment:
I think you are right, there may be a problem in this because an existing returned checkout will always match.
However, a search like this, while working, is not scalable in the real world, because you end up loading your whole database in a Python process.